### PR TITLE
PullRequest for Issue80: add the feature to support the timeDelta for server app

### PR DESCRIPTION
**Pull request by XuetaoLiu-S**
As described in Issue80,

On client side:
   when creating the AMDSClientRequest, we will add the current local time (in UTC) to the message

On server side:
   when receiving the clientRequest, server will calculate the timeDelta;
   For the time based clientDataRequest, server will add the timeDelta to the time automatically and use the new "altered" time to populate the data

### CODE REVIEW

File: source/ClientRequest/AMDSClientRequest.cpp [hunk 1]
```diff
@@ -93,6 +94,14 @@ void AMDSClientRequest::printData()
 	AMErrorMon::information(this, AMDS_CLIENTREQUEST_INFO_REQUEST_DATA, toString());
 }
 
+int AMDSClientRequest::calculateTimeDelta()
```
Comment (dretrex) [inline source/ClientRequest/AMDSClientRequest.cpp]: This would be a `const` function.

File: source/DataElement/AMDSBufferGroup.cpp [hunk 2]
```diff
@@ -84,9 +84,9 @@ void AMDSBufferGroup::append(AMDSDataHolder *newData, bool elapsedDwellTime)
 		if (specturalCumulativeDataHolder) {
 			AMDSDwellStatusData cumulativeStatusData = specturalCumulativeDataHolder->dwellStatusData();
 
-			emit continuousDataUpdate(specturalCumulativeDataHolder);
-			emit continuousStatusDataUpdate(cumulativeStatusData, count());
-			emit continuousAllDataUpdate(specturalCumulativeDataHolder, cumulativeStatusData, count(), elapsedDwellTime);
+//			emit continuousDataUpdate(specturalCumulativeDataHolder);
```
Comment (dretrex) [inline source/DataElement/AMDSBufferGroup.cpp:87]: Here are some commented out lines.  Are we wanting to keep these?

File: source/DataElement/AMDSBufferGroup.h [hunk 3]
```diff
@@ -76,6 +76,8 @@ public slots:
 	void dwellFinishedAllDataUpdate(AMDSDataHolder *accumlatedDataHolder, AMDSDwellStatusData statusData, int count, double elapsedTime);
 
 protected:
+	/// calculate the delta between the client time and the server time i(in ms)
+	int calculateTimeDelta(AMDSClientRequest *clientRequest);
```
Comment (dretrex) [inline source/DataElement/AMDSBufferGroup.h:80]: This is likely a `const` function too.

File: source/Detector/Amptek/AmptekSDD123Detector.cpp [hunk 4]
```diff
@@ -172,7 +177,7 @@ void AmptekSDD123Detector::readStatusData(const QByteArray &statusData){
 
 	backwardsMid(8, 4, statusData, tmpData);
 	internalSetGeneralPurposeCounter(tmpData.toHex().toInt(&ok, 16));
-	//qDebug() << "gpCounter: " << tmpData.toHex() << gpCounter;
```
Comment (dretrex) [inline source/Detector/Amptek/AmptekSDD123Detector.cpp:175]: Some commented out debugs.

File: source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.cpp [hunk 5]
```diff
@@ -339,6 +340,7 @@ void AmptekSDD123EPICSDetectorManager::onClearSpectrumControlValueChange(double
 }
 
 void AmptekSDD123EPICSDetectorManager::onDwellTimeControlValueChange(double newValue){
+	qDebug() << "Got a value changed on dwellTime control as " << newValue;
```
Comment (dretrex) [inline source/Detector/Amptek/SGM/AmptekSDD123EPICSDetectorManager.cpp:343]: I'm assuming we're keeping these debugs around for a little while?

File: source/appController/AMDSAppController.cpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#include "AMDSAppController.h"
+
+#include "util/AMErrorMonitor.h"
+
+AMDSAppController *AMDSAppController::appController_ = 0;
+
+AMDSAppController *AMDSAppController::appController()
+{
+	if (!appController_)
+		AMErrorMon::error(0, 0, "The app controller is NOT instanced!");
```
Comment (dretrex) [inline source/appController/AMDSAppController.cpp:10]: Are we not wanting to do anything else?  Seems like it throws up the error but continues anyway.

File: source/appController/AMDSAppController.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+		Server = 1
+	};
+
+	static AMDSAppController *appController();
+	static void releaseAppController();
+
+	~AMDSAppController() {}
+
+	/// returns the current type of the app controller
+	inline AMDSAppController::AMDSAppType appType() { return appType_; }
```
Comment (dretrex) [inline source/appController/AMDSAppController.h]: `const` function.

**Comment by dretrex** [comment]
Looks good overall aside from the couple comments.
Time Estimate: 10 minutes

File: source/DataElement/AMDSBufferGroup.cpp (hunk 2 above)
Comment (XuetaoLiu-S) [inline source/DataElement/AMDSBufferGroup.cpp:87]: Yes. This will be finalized later.

File: source/Detector/Amptek/AmptekSDD123Detector.cpp (hunk 4 above)
Comment (XuetaoLiu-S) [inline source/Detector/Amptek/AmptekSDD123Detector.cpp:175]: We might want to keep the debugging messages for a little while and clean those up when we are "done"

File: source/appController/AMDSAppController.cpp (hunk 6 above)
Comment (XuetaoLiu-S) [inline source/appController/AMDSAppController.cpp:10]: This is a super class, the instance should be initialized by AMDSClientAppController / AMDSMainAppController. It works like AMBeamline

Comment (dretrex) [inline source/appController/AMDSAppController.cpp:10]: Ah, okay.

**Pull request merged**